flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1314,13 +1314,12 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
             if (!props.isSavepoint()) {
                 lastSubsumed =
                         addCompletedCheckpointToStoreAndSubsumeOldest(
-                                checkpointId,
-                                completedCheckpoint,
-                                pendingCheckpoint.getCheckpointPlan().getTasksToCommitTo());
+                                checkpointId, completedCheckpoint, pendingCheckpoint);
             } else {
                 lastSubsumed = null;
             }
 
+            pendingCheckpoint.getCompletionFuture().complete(completedCheckpoint);
```

Comment:
I have a small concern here current `CompleteExceptionally` is not in the same level with the complete. In this case, there is risk that if we change logic here, there might be case that the future is not done finally. 

To not change the order of "complete future" -> "report the checkpoint", we might also try marking the checkpoint in the catch block in this method.